R/insert-ui.R

```diff
@@ -172,3 +172,275 @@ removeUI <- function(selector,
   if (!immediate) session$onFlushed(callback, once = TRUE)
   else callback()
 }
+
+
+#' Dynamically add a tabPanel
+#'
+#' Dynamically add a \code{tabPanel()} into an existing \code{tabsetPanel}.
+#'
+#' @param tabsetPanelId The \code{id} of the \code{tabsetPanel()} into which
+#'   \code{tab} will be inserted.
+#'
+#' @param tab The tab element to be added (must be created with
+#'   \code{tabPanel}).
+#'
+#' @param target The \code{value} of an existing \code{tabPanel()}, next to
+#'   which \code{tab} will be added. If \code{NULL},the \code{tab} will be
+#'   placed either as the first tab or the last tab, depending on the
+#'   \code{position} argument.
+#'
+#' @param position Should \code{tab} be added to the right ot to the left
+#'   of \code{target}?
+#'
+#' @param immediate Whether \code{tab} should be immediately inserted into
+#'   the app when you call \code{insertTab}, or whether Shiny should wait until
+#'   all outputs have been updated and all observers have been run (default).
+#'
+#' @param session The shiny session within which to call \code{insertTab}.
+#'
+#' @seealso \code{\link{removeTab}}, \code{\link{showTab}},
+#'   \code{\link{hideTab}}
+#'
+#' @examples
+#' ## Only run this example in interactive R sessions
+#' if (interactive()) {
+#' ui <- fluidPage(
+#'   sidebarLayout(
+#'     sidebarPanel(actionButton("add", "Add tab")),
+#'     mainPanel(
+#'       tabsetPanel(id = "tabs",
+#'         tabPanel("Hello", "This is the hello tab"),
+#'         tabPanel("Foo", "This is the foo tab"),
+#'         tabPanel("Bar", "This is the bar tab")
+#'       )
+#'     )
+#'   )
+#' )
+#' server <- function(input, output, session) {
+#'   observeEvent(input$add, {
+#'     insertTab(tabsetPanelId = "tabs",
+#'       tabPanel("Dynamic", "This a dynamically-added tab"),
+#'       target = "Bar"
+#'     )
+#'   })
+#' }
+#'
+#' shinyApp(ui, server)
+#' }
+#' @export
+insertTab <- function(tabsetPanelId, tab, target = NULL,
+  position = c("right", "left"), immediate = FALSE,
+  session = getDefaultReactiveDomain()) {
+
+  force(tabsetPanelId)
+  force(tab)
+  force(target)
+  position <- match.arg(position)
+  force(session)
+
+  iconClass <- tab$attribs$`data-icon-class`
+  icon <- if (!is.null(iconClass)) {
+    # for font-awesome we specify fixed-width
+    if (grepl("fa-", iconClass, fixed = TRUE))
+      iconClass <- paste(iconClass, "fa-fw")
+    icon(name = NULL, class = iconClass)
+  } else NULL
+
+  callback <- function() {
+    session$sendInsertTab(
+      tabsetPanelId = tabsetPanelId,
+      tab = processDeps(tab, session),
+      icon = processDeps(icon, session),
+      target = target,
+      position = position)
+  }
+
+  if (!immediate) session$onFlushed(callback, once = TRUE)
+  else callback()
+}
+
+
+#' Dynamically remove a tabPanel
```

Comment:
I struggle a bit with this. I think it definitely makes sense to link the functions around (maybe even have the same description at the top, with links to the docs of four all functions right there). So, why not put it all in a single help file, like you suggest? My thinking was that I, as an user, would much prefer to have one help file per function, since that allows me to have a usage example for that function only. Since these are related, but different, functions, seeing a usage example for each one could be beneficial (or at least that's my theory). I'm really not sure though, so I'd love to know your and @jcheng5's opinion about this, since I imagine it's common "problem" when writing documentation for family of related, but different, functions...